### PR TITLE
Support empty input tensor for some ops (fix #14657)

### LINKED ISSUE #14657: FusedBatchNorm & Conv2D backwards doesn't support zero batch size
Most ops in TF work well with tensors with zero elements. However, <del>convolution</del> fusedbatchnorm with cudnn gives the following error:
```
2017-11-17 08:00:20.835113: F tensorflow/stream_executor/cuda/cuda_dnn.cc:444] could not convert BatchDescriptor {count: 0 feature_map_count: 1024 spatial: 28 28  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX} to cudnn tensor descriptor: CUDNN_STATUS_BAD_PARAM 
```

I would expect it checks and returns a 4D tensor with zero batch-size. Currently I have to work around it by `tf.cond`.

### PR DESCRIPTION
**Pull request by ppwwyyxx**
Cudnn kernels doesn't work for empty input tensors.
This PR adds support for empty input tensor for FusedBatchNorm,FusedBatchNormGrad,Conv2DBackpropFilter, and cudnn pooling. (fix #14657)

### CODE REVIEW

File: tensorflow/core/kernels/conv_grad_filter_ops.cc [hunk 1]
```diff
@@ -595,6 +596,12 @@ class Conv2DSlowBackpropFilterOp : public OpKernel {
     if (filter_shape.num_elements() == 0) {
       return;
     }
+    if (input.shape().num_elements() == 0) {
```
Comment (yzhwang) [inline tensorflow/core/kernels/conv_grad_filter_ops.cc]: Could you also add this to conv_input_filter_ops.cc?
Also, could you make the comments for this consistent by saying something like: if there is nothing to cmpute, return empty tensors as the output.

**Review by yzhwang — CHANGES_REQUESTED** [review]
Thanks for the PR!
Could you add according test cases for empty input tensor  for all these ops?
fused_batch_norm test is at here:https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_fused_batchnorm_test.py
Others are located at tensorflow/python/kernel_tests.

File: tensorflow/core/kernels/conv_grad_filter_ops.cc (hunk 1 above)
Comment (ppwwyyxx) [reply tensorflow/core/kernels/conv_grad_filter_ops.cc]: You mean `conv_grad_filter_ops.cc`? It has handled this correctly.
I'll add comments.

**Comment by ppwwyyxx** [comment]
When input tensor has zero elements, the reference BN forward implementation in the test script gives NaN for mean/variance, however I return zeros.
NaN seems to make more sense in terms of math, though it will then require some special treatment in the actual training. If there is no objections I'll switch to NaNs.

**Comment by yzhwang** [comment]
@zhangyaobit Could you comment on this: https://github.com/tensorflow/tensorflow/pull/15264#issuecomment-352926246 please?

**Review by yzhwang — APPROVED** [review]
LGTM.

**Comment by yzhwang** [comment]
@tensorflow-jenkins test this please

**Comment by ppwwyyxx** [comment]
I haven't switched from returning zeros to returning NaNs, so the test is failing.

**Comment by yzhwang** [comment]
Let's wait for @zhangyaobit 's comment on this first then.

**Review by yzhwang — CHANGES_REQUESTED** [review]
Undo LGTM for now until fix fused_batch_norm test.

**Comment by zhangyaobit** [comment]
@ppwwyyxx, could you comment on the use case of an empty input (e.g. [0, 64, 64, 3])? Should we require the empty is non-empty?

**Comment by ppwwyyxx** [comment]
In object detection we may run CNN on patches (object candidates) cropped from an image with predicted boxes. When the image has no candidates we'll get zero patches. In training we can filter out these data but still can't avoid it in testing. A workaround is to use `tf.cond` but I hope the op can support it by itself.

In fact, I just found that the CPU (eigen) implementation of `fused_batch_norm` can work with empty input and returns NaNs for mean/variance.

**Comment by zhangyaobit** [comment]
Thanks! This sounds good. Could you the behavior of GPU implementation consistent with Eigen (return NaNs)? Please let me know once the PR is ready for review.

**Comment by ppwwyyxx** [comment]
@zhangyaobit The changes have been made. Could you review it when you have a time? Thanks!

File: tensorflow/core/kernels/fused_batch_norm_op.cc [hunk 2]
```diff
@@ -239,6 +240,14 @@ struct FusedBatchNorm<GPUDevice, T, U> {
             << " offset shape: " << offset.shape().DebugString()
             << " tensor format: " << tensor_format;
 
+    // If input is empty, weturn NaN mean/variance
```
Comment (yzhwang) [inline tensorflow/core/kernels/fused_batch_norm_op.cc]: s/weturn/return.

**Review by yzhwang — COMMENTED** [review]
LGTM.

**Review by zhangyaobit — APPROVED** [review]
Thanks!

**Comment by zhangyaobit** [comment]
@tensorflow-jenkins test this please

**Review by yzhwang — APPROVED** [review]

**Comment by ppwwyyxx** [comment]
The implementation of `FillFunctor`, `SetZeroFunctor`, etc, are split in two bazel targets: `:fill_functor` and `:constant_op`. However `:constant_op` depends on a lot of stuff: it depends on `:transpose_functor` which depends on `conv2d` (I saw a TODO for this by @yzhwang). But I need `conv2d` to use `SetZeroFunctor` which ends up being a cyclic reference.

**Comment by ppwwyyxx** [comment]
```diff
-    deps = ARRAY_DEPS,
+    deps = [
+        "//tensorflow/core:array_grad",
+        "//tensorflow/core:array_ops_op_lib",
+        "//tensorflow/core:framework",
+        "//tensorflow/core:lib",
+        "//third_party/eigen3",
+        ":bounds_check",
+        ":fill_functor",
+        ":ops_util",
+    ],
```
Cherry-picking the dependencies seems to make this PR build, but doesn't sound like an ideal solution. In general I guess functors should not depend on ops, but here `fill_functor` is actually in `:constant_op`, and `:transpose_functor` depends on `:conv_ops`.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by ppwwyyxx** [comment]
Thanks @drpngx for help! However I haven't yet fixed the cyclic dependency error mentioned above. I 
think a proper fix would be one of the following:
1. Move GPU implementation of `fill_functor` to target `:fill_functor`.
2. Don't let `:transpose_functor` depend on `:conv_ops`.

The first one seems to be within my reach. I can give it a try but not sure if there is any reason why this is not done before.

**Comment by drpngx** [comment]
Yeah, I would go for moving the stuff out of the constant op to the fill functor. If you feel that it is a generic library that could be used by more than one, then you could extract it out of constant op and have both depend on it.

**Comment by drpngx** [comment]
@yifeif for some reason, this is stuck with `kokoro-run`. There are others PRs as well.

**Comment by yifeif** [comment]
Ah if a PR has ran Kokoro tests before, it will need the force-run tag :).

**Comment by drpngx** [comment]
Oh, makes sense, of course.

**Comment by drpngx** [comment]
@ppwwyyxx there are some build breakages on GPU, could you check?

**Comment by ppwwyyxx** [comment]
The build was failing because of the bazel dependency problem, which should've been fixed now after I moved implementations to `:fill_functor` target.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
/CC @gunan ran out of devmapper space on the Jenkins build.
```
devmapper: Thin Pool has 968455 free data blocks which is less than minimum required 983040 free data blocks. Create more free space in thin pool or use dm.min_free_space option to change behavior
ERROR: docker build failed. Dockerfile is at /var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/tensorflow/tools/ci_build/Dockerfile.cpu
```

Jenkins, test this please.

**Comment by ppwwyyxx** [comment]
Why is 'XLA' build failing without details?

**Comment by drpngx** [comment]
OK, looks like @yifeif might have fixed it. We just ran out of space on the machine.

**Comment by yifeif** [comment]
@ppwwyyxx looks like an internal infra failure. I kicked it off again.

**Comment by drpngx** [comment]
Merged. Woohoo!

**Pull request merged**